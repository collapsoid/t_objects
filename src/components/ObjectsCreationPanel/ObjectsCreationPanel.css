.objects-creation-panel {
  background-color: #fefefe;
  padding: 15px;
  border-radius: 2px;
}

.objects-creation-panel__header {
  display: flex;
  justify-content: space-between;
  padding: 5px;
}

.objects-creation-panel__title {
  white-space: nowrap;
  font-weight: bold;
}

.objects-creation-panel__summary {
  padding-top: 25px;
}

.objects-creation-panel__total-quantity {
  display: inline-block;
  padding-left: 40px;
}

.objects-creation-panel__field-group {
  padding-top: 15px;
}

.field-group {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
}

.field-group div {
  margin-right: 5px;
}

.object-name input {
  width: 100%;
  padding: 15px 5px;
  box-sizing: border-box;
}