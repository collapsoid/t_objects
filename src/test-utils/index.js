import React from 'react';
import {createStore, combineReducers} from 'redux';
import {Provider} from 'react-redux';
import {render as rtlRender} from '@testing-library/react';

import * as reducers from '../store/reducers';

function render(
  ui,
  {
    initialState,
    store = createStore(combineReducers({...reducers}), initialState),
    ...renderOptions
  } = {}) {
  function Wrapper({children}) {
    return <Provider store={store}>{children}</Provider>;
  }

  return rtlRender(ui, {wrapper: Wrapper, ...renderOptions});
}

export * from '@testing-library/react';
export {render};