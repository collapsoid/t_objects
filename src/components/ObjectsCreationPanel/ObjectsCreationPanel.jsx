import React, {useRef, useState} from 'react';
import {connect} from 'react-redux';

import './ObjectsCreationPanel.css';

import {FormProcessor} from '../../utils';
import {addTargetObjects} from '../../store/actions';

import {withForm} from '../HOC';
import ObjectsCounter from '../ObjectsCounter/ObjectsCounter';


const ObjectsCreationPanel = withForm(({formRef, children, currentTargetObjects, addTargetObjects, setEditorState}) => {
  const planesRef = useRef(null);
  const objectsOnPlaneRef = useRef(null);
  const [objectsQuantity, setObjectsQuantity] = useState(0);

  const calculateObjectsQuantity = () => {
    setObjectsQuantity(+planesRef.current.value * +objectsOnPlaneRef.current.value);
  };

  const submitHandler = () => {
    formRef.current.requestSubmit();

    if (!formRef.current.formValid) {
      return;
    }

    const formProcessor = new FormProcessor();
    const objectsParameters = {...formProcessor.collectFormFieldsValue(formRef.current), objectsQuantity};

    addTargetObjects(objectsParameters);
    setEditorState(null);
  };

  const fieldProps = {
    planes: {
      ref: planesRef,
      onChange: calculateObjectsQuantity
    },
    objectsOnPlane: {
      ref: objectsOnPlaneRef,
      onChange: calculateObjectsQuantity
    }
  };
  return (
    <section className="objects-creation-panel">
      <header className="objects-creation-panel__header">
        <span className="objects-creation-panel__title">Создание регулярной группы целевых объектов</span>
      </header>

      {children(currentTargetObjects.template, fieldProps)}

      <div className="objects-creation-panel__summary">
        <button className="objects-creation-panel__create-group-btn form-btn" type="submit" onClick={submitHandler}>СОЗДАТЬ ГРУППУ ОБЪЕКТОВ</button>
        <div className="objects-creation-panel__total-quantity">
          <ObjectsCounter count={objectsQuantity} />
        </div>
      </div>
    </section>
  );
});

const mapState = ({currentTargetObjects}) => ({currentTargetObjects});

export default connect(mapState, {addTargetObjects})(ObjectsCreationPanel);