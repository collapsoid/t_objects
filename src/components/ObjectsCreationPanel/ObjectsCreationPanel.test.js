import React from 'react';
import {fireEvent, render} from '../../test-utils';

import ObjectsCreationPanel from './ObjectsCreationPanel';

describe('ObjectsCreationPanel', () => {
  const initialState = {
    currentTargetObjects: {
      template: []
    }
  };

  test('should render creation panel', () => {
    const {asFragment} = render(<ObjectsCreationPanel />, {initialState});
    const component = asFragment();
    
    expect(component).toMatchSnapshot();
  });

  test('should render an invalid field value error', () => {
    const {getByText} = render(<ObjectsCreationPanel />, {initialState});

    fireEvent.click(getByText(/создать группу объектов/i));

    expect(getByText(/введите/i)).toBeTruthy();
  });

  test('should reset the editor state after successful validation', () => {
    const setEditorStateMock = jest.fn();
    const {getByPlaceholderText, getByText} = render(
      <ObjectsCreationPanel setEditorState={setEditorStateMock} />,
      {initialState}
    );

    fireEvent.change(getByPlaceholderText(/название объекта/i), {target: {value: 'test value'}});
    fireEvent.click(getByText(/создать группу объектов/i));

    expect(setEditorStateMock).toBeCalledWith(null);
  });
});