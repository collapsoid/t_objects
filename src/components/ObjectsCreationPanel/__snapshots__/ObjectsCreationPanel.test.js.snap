// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectsCreationPanel should render creation panel 1`] = `
<DocumentFragment>
  <section
    class="objects-creation-panel"
  >
    <header
      class="objects-creation-panel__header"
    >
      <span
        class="objects-creation-panel__title"
      >
        Создание регулярной группы целевых объектов
      </span>
    </header>
    <form
      novalidate=""
    >
      <div
        class="object-name"
      >
        <div
          class="objects-form-field"
        >
          <label
            class="objects-form-field__label"
          >
            ЗАДАЙТЕ НАЗВАНИЕ ОБЪЕКТА
          </label>
          <div
            class="objects-form-field__container"
          >
            <input
              class="objects-form-field__input"
              data-type="text"
              name="title"
              placeholder="Название объекта"
              required=""
              type="text"
            />
            <span
              class="objects-form-field__error"
            >
              Введите название объекта
            </span>
          </div>
        </div>
      </div>
    </form>
    <div
      class="objects-creation-panel__summary"
    >
      <button
        class="objects-creation-panel__create-group-btn form-btn"
        type="submit"
      >
        СОЗДАТЬ ГРУППУ ОБЪЕКТОВ
      </button>
    </div>
  </section>
</DocumentFragment>
`;
